### PR TITLE
Implement set segment state call from RootCoord to DataCoord.

### PR DESCRIPTION
**Pull request by soothing-rain**
It is called when RootCoord believes that an import task's segments are ready to go online (become searchable). It is done by flipping the states of the segments from `importing` to `flushed`.

/kind feature

issue: #15604
Signed-off-by: Yuchen Gao <yuchen.gao@zilliz.com>

### CODE REVIEW

File: internal/datacoord/services.go [hunk 1]
```diff
@@ -446,6 +446,32 @@ func (s *Server) DropVirtualChannel(ctx context.Context, req *datapb.DropVirtual
...
+	}
+	err := s.meta.SetState(req.GetSegmentId(), req.GetNewState())
+	if err != nil {
+		return &datapb.SetSegmentStateResponse{
+			Status: &commonpb.Status{
+				ErrorCode: commonpb.ErrorCode_UnexpectedError,
+				Reason:    err.Error(),
+			},
+		}, nil
+	}
```
Comment (congqixia) [inline internal/datacoord/services.go:467]: log is needed in error handling branch

File: internal/proto/data_coord.proto [hunk 2]
```diff
@@ -397,6 +398,15 @@ message WatchChannelsResponse {
   common.Status status = 1;
 }
 
+message SetSegmentStateRequest {
+  int64 segment_id = 1;
```
Comment (congqixia) [inline internal/proto/data_coord.proto:402]: please add `common.MsgBase` as first member for a request message.

File: internal/proto/data_coord.proto [hunk 3]
```diff
@@ -397,6 +398,15 @@ message WatchChannelsResponse {
...
+  common.SegmentState new_state = 2;
+}
+
+message SetSegmentStateResponse {
+  common.Status status = 1;
```
Comment (congqixia) [inline internal/proto/data_coord.proto:407]: if `common.Status` is the only result to return, the rpc signature could just return `common.Status`

File: internal/rootcoord/import_manager.go [hunk 4]
```diff
@@ -476,7 +470,7 @@ func (m *importManager) expireOldTasksLoop(wg *sync.WaitGroup) {
 			return
 		case <-ticker.C:
 			log.Info("(in loop) starting expiring old tasks...",
-				zap.Any("cleaning up interval", time.Duration(Params.RootCoordCfg.ImportTaskExpiration)))
+				zap.Duration("cleaning up interval", time.Duration(Params.RootCoordCfg.ImportTaskExpiration)))
```
Comment (congqixia) [inline internal/rootcoord/import_manager.go:473]: maybe `Params.RootCoordCfg.ImportTaskExpiration * time.Second` ?

File: internal/rootcoord/root_coord.go [hunk 5]
```diff
@@ -714,6 +720,29 @@ func (c *Core) SetDataCoord(ctx context.Context, s types.DataCoord) error {
...
+		resp, err := s.SetSegmentState(ctx, req)
+		if err != nil || resp.GetStatus().GetErrorCode() != commonpb.ErrorCode_Success {
+			log.Error("failed to update segment state",
+				zap.Any("request", req), zap.Any("response", resp), zap.Error(err))
+			return err
+		}
+		log.Info("successfully set segment state",
+			zap.Any("segment ID", req.GetSegmentId()),
+			zap.Any("new segment state", req.GetNewState()))
+		return nil
```
Comment (congqixia) [inline internal/rootcoord/root_coord.go:743]: NIT: Int64 for SegmentID, String for State.String()?

File: internal/rootcoord/root_coord.go [hunk 6]
```diff
@@ -2428,30 +2458,40 @@ func (c *Core) checkCompleteIndexLoop(ctx context.Context, ti *datapb.ImportTask
...
-// segmentsOnlineReady returns true if segments are ready to go up online (a.k.a. searchable).
+// bringSegmentsOnline brings the segments online so that data in these segments become searchable
+// it is done by changing segments' states from `importing` to `flushed`.
+func (c *Core) bringSegmentsOnline(ctx context.Context, segIDs []UniqueID) {
+	log.Info("bringing import task's segments online!", zap.Any("segment IDs", segIDs))
+	for _, id := range segIDs {
+		// Explicitly mark segment states `flushed`.
+		c.CallUpdateSegmentStateService(ctx, id, commonpb.SegmentState_Flushed)
+	}
+}
```
Comment (congqixia) [inline internal/rootcoord/root_coord.go:2492]: Looks like this operation is not atomic.

File: internal/datacoord/services.go (hunk 1 above)
Comment (soothing-rain) [reply internal/datacoord/services.go:467]: DONE.

File: internal/proto/data_coord.proto (hunk 2 above)
Comment (soothing-rain) [reply internal/proto/data_coord.proto:402]: DONE.

File: internal/proto/data_coord.proto (hunk 3 above)
Comment (soothing-rain) [reply internal/proto/data_coord.proto:407]: I meant to reserve this response type to support additional info. For example if we want to make this atomic we definitely need more info :)

File: internal/rootcoord/import_manager.go (hunk 4 above)
Comment (soothing-rain) [reply internal/rootcoord/import_manager.go:473]: Done.

File: internal/rootcoord/root_coord.go (hunk 5 above)
Comment (soothing-rain) [reply internal/rootcoord/root_coord.go:743]: Done.

File: internal/rootcoord/root_coord.go (hunk 6 above)
Comment (soothing-rain) [reply internal/rootcoord/root_coord.go:2492]: Yes it should be atomic. I left a TODO here and will address it in the next PR.

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**